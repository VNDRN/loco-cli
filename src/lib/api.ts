import fetch from "isomorphic-unfetch";
import { Translations } from "../types";

const fetchApi = async <T>(path: string, opts = {}) => {
  const response = await fetch(`https://localise.biz/api${path}`, opts);
  if (!response.ok) {
    throw new Error(`HTTPError: ${response.status} ${response.statusText}`);
  }
  const json = (await response.json()) as T;
  return json;
};

export const apiPull = (key: string) =>
  fetchApi<Translations>(`/export/all.json?key=${key}`);

export const apiPush = (
  key: string,
  locale: string,
  translations: Translations[string]
) =>
  fetchApi<void>(`/import/json?key=${key}&locale=${locale}`, {
    method: "POST",
    body: JSON.stringify(translations),
  });
