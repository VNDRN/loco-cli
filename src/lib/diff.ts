import { detailedDiff } from "deep-object-diff";
import { Translations } from "../types";
import { dotObject } from "./dotObject";

interface DetailedDiff {
  added: object;
  updated: object;
  deleted: object;
}

export const diff = (local: Translations, remote: Translations) => {
  const { added, updated, deleted } = detailedDiff(
    local,
    remote
  ) as DetailedDiff;

  return {
    added: dotObject(added),
    updated: dotObject(updated),
    deleted: dotObject(deleted),
  };
};
