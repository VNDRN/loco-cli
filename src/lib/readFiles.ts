import { existsSync, readdirSync, stat, statSync } from "fs";
import { readFile } from "fs";
import { join } from "path";
import { Translations } from "../types";
import { exitError } from "../util/exit";

const readJSON = async (path: string) => {
  if (!existsSync(path)) {
    exitError(`File not found: ${path}`);
  }

  return new Promise((resolve, reject) => {
    readFile(path, "utf8", (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(JSON.parse(data));
    });
  }) as Promise<Record<string, string>>;
};

const readFilesInDir = async (path: string, separator?: string) => {
  const res = {};
  if (!existsSync(path)) {
    exitError(`Directory not found: "${path}"`);
  }
  const files = readdirSync(path);

  await Promise.all(
    files.map(async (file) => {
      if (file.endsWith(".json")) {
        const json = await readJSON(join(path, file));
        Object.keys(json).forEach((key) => {
          // @ts-expect-error
          res[`${file.replace(".json", "")}${separator}${key}`] = json[key];
        });
      }
    })
  );
  return res;
};

export const readFiles = async (
  dir: string,
  useNamespaces: boolean
): Promise<Translations> => {
  if (!existsSync(dir)) {
    throw new Error(`Directory not found: "${dir}"`);
  }
  const translations: Translations = {};
  const locales = readdirSync(dir);
  await Promise.all(
    locales.map(async (locale) => {
      const localeDir = join(dir, locale);
      if (useNamespaces) {
        // using namespaces; locales should be folders
        if (statSync(localeDir).isDirectory()) {
          translations[locale] = await readFilesInDir(localeDir, ":");
        }
      } else {
        // using namespaces; locales shoud be JSON files
        if (locale.endsWith(".json")) {
          translations[locale.replace(".json", "")] = await readJSON(
            join(dir, locale)
          );
        }
      }
    })
  );
  return translations;
};
