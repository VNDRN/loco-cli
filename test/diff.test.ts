import { diff } from "../src/lib/diff";
import { local, remote } from "./mock";

test("returns diff", () => {
  expect(diff(local, remote)).toMatchSnapshot();
});
