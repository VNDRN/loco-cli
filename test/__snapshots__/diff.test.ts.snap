// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns diff 1`] = `
Object {
  "added": Object {
    "en": Object {
      "body": "Newly added key remote in EN only",
      "buttons": Object {
        "accept": "Newly added key deep remote in EN only",
      },
    },
  },
  "deleted": Object {
    "en": Object {
      "buttons": Object {
        "skip": undefined,
      },
      "title": undefined,
    },
  },
  "updated": Object {
    "en": Object {
      "buttons": Object {
        "confirm": "Different translation local/remote",
      },
    },
  },
}
`;
